mypy/plugins/dataclasses.py

```diff
@@ -738,6 +739,33 @@ def _get_bool_arg(self, name: str, default: bool) -> bool:
             return require_bool_literal_argument(self._api, expression, name, default)
         return default
 
+    def _get_default_init_value_for_field_specifier(self, call: Expression) -> bool:
+        """
+        Find a default value for the `init` parameter of the specifier being called. If the
+        specifier's type signature includes an `init` parameter with a type of `Literal[True]` or
+        `Literal[False]`, return the appropriate boolean value from the literal. Otherwise,
+        fall back to the standard default of `True`.
+        """
+        if not isinstance(call, CallExpr):
+            return True
+
+        specifier_type = _get_callee_type(call)
```

Comment:
I'm reusing this existing helper from [`mypy.plugins.common`](https://github.com/python/mypy/blob/master/mypy/plugins/common.py#L90), but unfortunately it doesn't truly support overloads:

```
    if isinstance(callee_node, (Var, SYMBOL_FUNCBASE_TYPES)) and callee_node.type:
        callee_node_type = get_proper_type(callee_node.type)
        if isinstance(callee_node_type, Overloaded):
            # We take the last overload.
            return callee_node_type.items[-1]
        elif isinstance(callee_node_type, CallableType):
            return callee_node_type
```

I'm not sure if there's an existing method that can resolve the appropriate overload or if some new plumbing will need to be added to MyPy.

Reply:
We can't generally resolve overloads in a dataclass plugin, since semantic analysis is not complete yet, and certain type operations aren't strictly speaking available yet. We could perhaps do a best-effort resolution, but I'm not sure if it's worth the effort. Do you know of any use case where this is needed? Perhaps we can come up with a simplified implementation that is good enough for common use cases.

Reply:
@JukkaL I'm not sure if anything is using it in the wild, thought he PEP calls it out with an explicit example: https://peps.python.org/pep-0681/#field-specifier-parameters We've discussed this some in https://github.com/python/mypy/issues/14293#issuecomment-1455150506, so I can try asking there if there any concrete use cases to be aware of.